observability/src/main/java/io/grpc/observability/logging/GcpLogSink.java

```diff
@@ -0,0 +1,155 @@
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Sink for Google Cloud Logging.
+ */
+public class GcpLogSink implements Sink {
+
+  private static final Logger logger = Logger.getLogger(Sink.class.getName());
+
+  private static final String SERVICE_TO_EXCLUDE = "google.logging.v2.LoggingServiceV2";
+  private static final String DEFAULT_LOG_NAME = "grpc";
+
+  private static GcpLogSink gcpLogSinkInstance = null;
+  private final LoggingOptions gcpLoggingOptions;
+  private final Logging gcpLoggingClient;
+
+  /**
+   * Constructor for custom sink.
+   *
+   * @param destinationProjectId cloud project id to write logs
+   */
+  private GcpLogSink(String destinationProjectId) {
+    if (Strings.isNullOrEmpty(destinationProjectId)) {
+      gcpLoggingOptions = LoggingOptions.getDefaultInstance();
+    } else {
+      gcpLoggingOptions = LoggingOptions.newBuilder().setProjectId(destinationProjectId).build();
+    }
+    gcpLoggingClient = gcpLoggingOptions.getService();
+  }
+
+  /**
+   * Initialize GcpLogSInk.
+   *
+   * @param projectId cloud project id to write logs
+   */
+  // TODO(dnvindhya) read the projectId value from config instead of taking
+  // it as an argument
+  public static synchronized GcpLogSink init(String projectId) {
+    if (gcpLogSinkInstance != null) {
+      throw new IllegalStateException("GcpLogSink already initialized");
+    }
+    gcpLogSinkInstance = new GcpLogSink(projectId);
+    return gcpLogSinkInstance;
+  }
+
+  /**
+   * Retrieves a single instance of GcpLogSink.
+   */
+  public static GcpLogSink getInstance() {
+    if (gcpLogSinkInstance == null) {
+      throw new IllegalStateException("GcpLogSink not initialized");
+    }
+    return gcpLogSinkInstance;
+  }
+
+  @Override
+  public synchronized void write(GrpcLogRecord logProto) {
+    if (gcpLoggingClient == null) {
+      logger.log(Level.SEVERE, "Attempt to write after GcpLogSink is closed.");
+      return;
+    }
+    if (SERVICE_TO_EXCLUDE.equals(logProto.getServiceName())) {
+      return;
+    }
+    try {
+      Map<String, Object> mapPayload = protoToMapConverter(logProto);
+      Severity logEntrySeverity = getCloudLoggingLevel(logProto.getLogLevel());
+      // TODO(vindhyan): make sure all (int, long) values are not displayed as double
+      LogEntry grpcLogEntry =
+          LogEntry.newBuilder(JsonPayload.of(mapPayload))
+              .setSeverity(logEntrySeverity)
+              .setLogName(DEFAULT_LOG_NAME)
+              .setResource(MonitoredResource.newBuilder("global").build())
+              .build();
+      gcpLoggingClient.write(Collections.singleton(grpcLogEntry));
+    } catch (Exception e) {
+      logger.log(Level.SEVERE, "Caught exception while writing to cloud logging", e);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private Map<String, Object> protoToMapConverter(GrpcLogRecord logProto)
+      throws InvalidProtocolBufferException, IOException {
```

Comment:
"There is a more general exception, 'java.io.IOException', in the throws list already. " So you can just remove `InvalidProtocolBufferException`

Reply:
Updated